cmsplugin_zinnia/cms_toolbar.py

```diff
@@ -9,25 +9,42 @@
 class ZinniaToolbar(CMSToolbar):
 
     def populate(self):
+        user = self.request.user
         zinnia_menu = self.toolbar.get_or_create_menu(
             'zinnia-menu', _('Zinnia'))
 
         url = reverse('admin:zinnia_entry_add')
-        zinnia_menu.add_sideframe_item(_('New entry'), url=url)
+        zinnia_menu.add_sideframe_item(
+            _('New entry'), url=url,
+            disabled=not user.has_perm('zinnia.add_entry'))
 
         url = reverse('admin:zinnia_category_add')
-        zinnia_menu.add_sideframe_item(_('New category'), url=url)
+        zinnia_menu.add_sideframe_item(
+            _('New category'), url=url,
+            disabled=not user.has_perm('zinnia.add_category'))
 
         zinnia_menu.add_break()
 
         url = reverse('admin:zinnia_entry_changelist')
-        zinnia_menu.add_sideframe_item(_('Entries list'), url=url)
+        zinnia_menu.add_sideframe_item(
+            _('Entries list'), url=url,
+            disabled=not user.has_perm('zinnia.change_entry'))
 
         url = reverse('admin:zinnia_category_changelist')
-        zinnia_menu.add_sideframe_item(_('Categories list'), url=url)
+        zinnia_menu.add_sideframe_item(
+            _('Categories list'), url=url,
+            disabled=not user.has_perm('zinnia.change_category'))
 
         url = reverse('admin:tagging_tag_changelist')
-        zinnia_menu.add_sideframe_item(_('Tags list'), url=url)
+        zinnia_menu.add_sideframe_item(
+            _('Tags list'), url=url,
+            disabled=not user.has_perm('tagging.change_tag'))
+
+        # remove complete menu if all items are disabled
+        for item in zinnia_menu.get_items():
+            if hasattr(item, 'disabled') and not item.disabled:
+                return
```

Comment:
The idea is actually easy:

> If in the (stupid) for-loop we detect an _enabled_ item then
> refrain from removing the toolbar menu.

It's only a bit difficult to understand, because the **for-loop is lengthy** (for the simple task it's supposed to achieve) and because the **logic is inverse**: It should better say

> If (and only if) all items are disabled then
> remove the toolbar menu.

We could only achieve this via a filter over the menu items. But that requires extra code (the filter).
``` python
if len(filter(enabled_items, zinnia_menu.get_items())) == 0:
    self.toolbar.remove_item(zinnia_menu)
```
Then again, if this is more readable ... ?